src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java

```diff
@@ -0,0 +1,45 @@
+package uk.gov.pay.directdebit.mandate.services.gocardless;
+
+import uk.gov.pay.directdebit.events.dao.GoCardlessEventDao;
+import uk.gov.pay.directdebit.gatewayaccounts.model.GoCardlessOrganisationId;
+import uk.gov.pay.directdebit.mandate.model.GoCardlessMandateId;
+import uk.gov.pay.directdebit.mandate.model.MandateState;
+import uk.gov.pay.directdebit.payments.model.GoCardlessEvent;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static uk.gov.pay.directdebit.mandate.model.MandateState.ACTIVE;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.CANCELLED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.CREATED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.FAILED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.SUBMITTED;
+
+public class GoCardlessMandateStateCalculator {
+
+    private final GoCardlessEventDao goCardlessEventDao;
+
+    private static final Map<String, MandateState> GOCARDLESS_ACTION_TO_MANDATE_STATE = Map.of(
```

Comment:
Wondering if this might work as a separate enum since its like a data type with each value having `GoCardlessAction` and `MandateAction`. I think this would mean it needn't be static on this class to make it accessible elsewhere and `GOCARDLESS_ACTIONS_THAT_CHANGE_STATE` just becomes the enum values and again needn't be declared static here. Also line 42 `.map(GOCARDLESS_ACTION_TO_MANDATE_STATE::get)` would become something like 
`.map(<EnumName>::getMandateState)` 🤷‍♂

Reply:
Perhaps but I want to keep some separation from a GoCardless action and what we interpret that to mean. The same action could potentially mean different things (even represent different states) in different circumstances. I worry that using an enum might encourage the insidious idea that a GoCardless action is, to an extent, just a state.

(I think it would be desirable to use an enum instead of a string for representing the action though. The only reason I didn’t was because we already have one and it isn’t quite what we want. It could be repurposed though. I also worry that GoCardless might introduce other actions that would break the necessary string-to-enum conversion code we’d have. I might drop them an email to check they won’t do that.)